lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <maths/CTreeShapFeatureImportance.h>
+
+#include <algorithm>
+#include <tuple>
+
+namespace ml {
+namespace maths {
+
+using TRowItr = core::CDataFrame::TRowItr;
+
+CTreeShapFeatureImportance::TDoubleVec
+CTreeShapFeatureImportance::shap(core::CDataFrame& frame,
+                                 const CDataFrameCategoryEncoder& encoder,
+                                 std::size_t numberFeatures,
+                                 std::size_t offset) {
+    numberFeatures = (numberFeatures != -1) ? numberFeatures : frame.numberColumns();
+    std::vector<std::size_t> maxDepthVec;
+    maxDepthVec.reserve(m_Trees.size());
+    for (auto& tree : m_Trees) {
+        auto samplesPerNode =
+            CTreeShapFeatureImportance::samplesPerNode(tree, frame, encoder, 0);
+        std::size_t maxDepth =
+            CTreeShapFeatureImportance::updateNodeValues(tree, 0, samplesPerNode, 0);
+        maxDepthVec.push_back(maxDepth);
+        m_SamplesPerNode.emplace_back(std::move(samplesPerNode));
+    }
+
+    auto result = frame.writeColumns(
+        m_NumberThreads, 0, frame.numberRows(),
+        core::bindRetrievableState(
+            [&](auto& state, TRowItr beginRows, TRowItr endRows) {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    auto encodedRow{encoder.encode(*row)};
+                    auto& phiSum = state;
+                    for (int i = 0; i < m_Trees.size(); ++i) {
+                        //                        phi[frame.numberColumns()] += m_Trees[i][0].value();
+                        SPath path(maxDepthVec[i] + 1);
+                        CTreeShapFeatureImportance::shapRecursive(
+                            m_Trees[i], m_SamplesPerNode[i], encoder,
+                            encodedRow, path, 0, 1.0, 1.0, -1, offset, row);
+                        i += 0;
+                    }
+                    for (int j = 0; j < numberFeatures; ++j) {
+                        row->writeColumn(offset + j,
+                                         (*row)[offset + j] / m_Trees.size());
```

Comment:
Can we pass in the normalisation factor and write once inside `shapRecursive`? I found this line confusing until I saw you were first writing them inside `shapRecursive`.

Reply:
Done